test-harness/src/ghciwatch.rs

```diff
@@ -227,97 +343,108 @@ impl GhciWatch {
-        // DEBUG watchexec::watchexec: spawning subtask {subtask="keyboard"}
-        // DEBUG watchexec::fs: launching filesystem worker
-        // DEBUG watchexec::watchexec: spawning subtask {subtask="error_hook"}
-        // DEBUG watchexec::fs: creating new watcher {kind="Poll(100ms)"}
-        // DEBUG watchexec::signal: launching unix signal worker
-        // DEBUG watchexec::fs: applying changes to the watcher {to_drop="[]", to_watch="[WatchedPath(\"src\")]"}
-        //
-        // "launching filesystem worker" is tempting, but the phrasing implies the event is emitted
-        // _before_ the filesystem worker is started (hence it is not yet ready to notice file
-        // events). Therefore, we wait for "applying changes to the watcher".
-        self.assert_logged(
-            Matcher::message("applying changes to the watcher").in_module("watchexec::fs"),
+    ///
+    /// Returns immediately if `ghciwatch` has already become ready to receive file events in the
+    /// current checkpoint.
+    pub async fn wait_until_watcher_started(&mut self) -> miette::Result<()> {
+        self.wait_for_log_with_timeout(
+            BaseMatcher::watcher_started(),
+            Some(self.current_checkpoint()),
+            self.default_timeout,
         )
         .await
-        .wrap_err("watchexec filesystem worker didn't start in time")
+        .wrap_err("notify watcher didn't start in time")?;
+        Ok(())
     }
 
     /// Wait until `ghciwatch` completes its initial load and is ready to receive file events.
+    ///
+    /// Returns immediately if `ghciwatch` has already completed its inital load and become ready to
+    /// receive file events in the current checkpoint.
     pub async fn wait_until_ready(&mut self) -> miette::Result<()> {
-        self.wait_until_started().await?;
-        self.wait_until_watcher_started().await?;
+        self.wait_for_log_with_timeout(
+            BaseMatcher::ghci_started().and(BaseMatcher::watcher_started()),
+            Some(self.current_checkpoint()),
+            self.startup_timeout,
+        )
+        .await
+        .wrap_err("ghciwatch didn't start in time")?;
         Ok(())
     }
 
     /// Wait until `ghciwatch` reloads the `ghci` session due to changed modules.
     pub async fn wait_until_reload(&mut self) -> miette::Result<()> {
         // TODO: It would be nice to verify which modules are changed.
-        self.assert_logged("^Reloading ghci:\n").await.map(|_| ())
+        self.wait_for_log(BaseMatcher::ghci_reload())
+            .await
+            .map(|_| ())
     }
 
     /// Wait until `ghciwatch` adds new modules to the `ghci` session.
     pub async fn wait_until_add(&mut self) -> miette::Result<()> {
         // TODO: It would be nice to verify which modules are being added.
-        self.assert_logged("^Adding modules to ghci:\n")
-            .await
-            .map(|_| ())
+        self.wait_for_log(BaseMatcher::ghci_add()).await?;
+        Ok(())
     }
```

Comment:
Looks like you have two styles here, and you're almost consistent, but not 100%

1. `foo.await.map(|_| ())`
2. `foo.await?; Ok(())`

Reply:
Nice catch